### PR TITLE
TypeScript source conversion

### PR DESCRIPTION
**Pull request by aaronbeall**
My first attempt to convert source to TypeScript per https://github.com/vkurchatkin/koa-connect/issues/11

I made the middleware function `req` and `res` types `any` because I wasn't really sure what they are expected to be. Initially I tried making them `IncomingMessage` and `ServerResponse` but this is not, for example, what `connect-history-api-fallback` uses.

I added scripts `start` and `build` to run the TS compiler (`start` in `--watch`). I'm not sure if `build` should be called in `prepublish` or something?

### CODE REVIEW

**Comment by lourd** [comment]
Thank you for this PR @aaronbeall, and extraordinary patience in waiting for a review.

Please take a look at the additions I made. Namely, I specified some types a little more strictly and converted the tests to TypeScript as well. I wouldn't feel good about converting the source to TypeScript without having the tests exercise the types.

Regarding the `any` types for `req` and `res`, I'm pretty sure you were on the right path with them being `IncomingMessage` and `ServerResponse`... unless Connect/Express use something else? I went ahead and used those, I'd rather not leave any `any` 🙃 

I'm looking for an LGTM from you or someone who will actually use this before merging, as it has been a few years since I actually used this library.

File: index.js [hunk 1]
```diff
@@ -1,38 +1,40 @@
...
  */
 function koaConnect(connectMiddleware) {
-  const handler = connectMiddleware.length < 3
-    ? noCallbackHandler
-    : withCallbackHandler
-  return function koaConnect(ctx, next) {
-    return handler(ctx, connectMiddleware, next)
-  }
+    return function koaConnect(ctx, next) {
+        return hasNoCallback(connectMiddleware)
```
Comment (jonaskello) [inline index.js]: It seems the old version evaluated which handler to use once outside the returned function and stored it in a constant used in the returned function. However, if I am reading the new version correctly, it evaluates which handler to use inside the returned function. This means the condition for which handler to use will run each time the returned function is called which may result in slightly decreased performance.

Comment (lourd) [reply index.js]: Yeah, that's a great point. Made some changes to keep the old behavior.

**Comment by aaronbeall** [comment]
@lourd Changes look good to me!

**Comment by lourd** [comment]
Just added an integration test to double-check that this would actually help in a practical scenario. I'm not sure it does... It requires [casting `ctx.req` to `any`](https://github.com/vkurchatkin/koa-connect/pull/13/commits/a5e42727487c4dfcebacacd245b7fa3fd127c0ac#diff-7226e856d75f43c509d6d7d22b11a79cR128-R132) to read the `body` parameter; i.e. the added types from middlewares don't make it through to subsequent middlewares.

I'm still really unsure about adding these changes, as I have never personally used Koa or Express in a project written with TypeScript. It doesn't seem like this PR would improve the situation that much given the observed behavior. Someone please correct me or the code if that's not the case, but as it stands I don't think this should be merged.

**Comment by jonaskello** [comment]
Typescript requires types to be able to use the library at all. So the alternative is to add your own types which is what I am doing. Because I am lazy I just add it as any for everything. So for me I already have any everywhere for this module. So just having integrated types with any everywhere is a step forward for me because then I don't have to add that myself :-). I think the first step is just adding the types. Then it can of course be iterated on further in more PRs.

**Comment by aaronbeall** [comment]
@lourd My two cents: unfortunately I'm no longer actively working on the project where I was using `koa-connect`, but I can say that the types I had written did indeed use `any` for `req` and `res` because otherwise it would raise type errors in my use case, and I couldn't really sort out what types those should be. This is an example where dynamic objects in JS make things tricky. It's like doable in TS, I just am not familiar enough with all the ways this should work in real use cases. However even with `any` for `req` and `res` it was still better than no types at all, which led to compiler type errors due to missing module definition and "implicit any" errors.

**Comment by lourd** [comment]
Alright, I agree. Better than no types at all 👍

**Comment by lourd** [comment]
Published in version 2.1.0

**Pull request merged**